larpandoracontent/LArTwoDReco/HitWidthClusterMergingAlgorithm.cc

```diff
@@ -0,0 +1,503 @@
+      
+        if(currentStatus != STATUS_CODE_SUCCESS || testStatus != STATUS_CODE_SUCCESS)
+            return false;
+    }
+    else
+    {
+        currentClusterDirection = GetClusterDirection(currentFitParameters, currentFitParameters.GetHigherXExtrema());
+        testClusterDirection = GetClusterDirection(testFitParameters, testFitParameters.GetLowerXExtrema());
+    }
+
+    if(testFitParameters.GetLowerXExtrema().GetX() > (currentFitParameters.GetHigherXExtrema().GetX() + m_maxXMergeDistance))
+      return false;
+
+    if(testFitParameters.GetLowerXExtrema().GetZ() > (currentFitParameters.GetHigherXExtrema().GetZ() + m_maxZMergeDistance) || testFitParameters.GetLowerXExtrema().GetZ() < (currentFitParameters.GetHigherXExtrema().GetZ() - m_maxZMergeDistance))
+      return false;
+    
+    if(fabs(currentClusterDirection.GetCosOpeningAngle(testClusterDirection)) < m_maxMergeCosOpeningAngle)
+      return false;
+
+    
+    // check that the new direction is consistent with the old clusters
+    if(m_useSlidingLinearFit)
+    {
+        const CartesianPointVector currentConstituentHitPositionVector(LArHitWidthHelper::GetConstituentHitPositionVector(currentFitParameters.GetConstituentHitVector()));
+        const CartesianPointVector testConstituentHitPositionVector(LArHitWidthHelper::GetConstituentHitPositionVector(testFitParameters.GetConstituentHitVector()));
+
+        CartesianPointVector newConstituentHitPositionVector(currentConstituentHitPositionVector);
+        newConstituentHitPositionVector.insert(newConstituentHitPositionVector.end(), testConstituentHitPositionVector.begin(), testConstituentHitPositionVector.end());
+      
+        const float layerPitch(LArGeometryHelper::GetWireZPitch(this->GetPandora()));
+	
+        TwoDSlidingFitResult newFitResult(&newConstituentHitPositionVector, m_layerFitHalfWindow, layerPitch);
+	
+        CartesianVector newClusterDirectionCurrent(0,0,0);
+        CartesianVector newClusterDirectionTest(0,0,0);
+				       
+        StatusCode currentStatus(newFitResult.GetGlobalFitDirectionAtX(currentFitParameters.GetHigherXExtrema().GetX(), newClusterDirectionCurrent));
+        StatusCode testStatus(newFitResult.GetGlobalFitDirectionAtX(testFitParameters.GetLowerXExtrema().GetX(), newClusterDirectionTest));
+      
+        if(currentStatus != STATUS_CODE_SUCCESS || testStatus != STATUS_CODE_SUCCESS)
+            return false;
+
+        if(std::fabs(newClusterDirectionCurrent.GetCosOpeningAngle(currentClusterDirection)) < m_maxDirectionDeviationCosAngle || std::fabs(newClusterDirectionTest.GetCosOpeningAngle(testClusterDirection)) < m_maxDirectionDeviationCosAngle)
+            return false;
+        
+    }
+    else
+    {
+        const LArHitWidthHelper::ConstituentHitVector currentConstituentHitVector(currentFitParameters.GetConstituentHitVector());
```

Comment:
Do you intend to support both options for the long term, or will we just make e.g. the sliding linear fit option the default and remove the else block (which should be retrievable via the git history).

Reply:
I don't think the sliding fit implementation works very well in this case. I find that a lot of the time fits cannot be made to clusters that consist of one hit (even when the hit is broken up). So I was planning on using my least squared fitting method which is built with these cases in mind. But I can reinvestigate this is if the sliding fit is preferred.

Reply:
Please go with whichever implementation you think works best. My question was more about whether you wanted to maintain two different approaches, when probably only one will get used. It's OK to maintain the two approaches if you want, but worth thinking about this anyhow.